v2/relay/relay.go

```diff
@@ -0,0 +1,463 @@
+package relay
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	pbv2 "github.com/libp2p/go-libp2p-circuit/v2/pb"
+	"github.com/libp2p/go-libp2p-circuit/v2/util"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	logging "github.com/ipfs/go-log"
+	pool "github.com/libp2p/go-buffer-pool"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+const (
+	ProtoIDv2Hop  = "/libp2p/circuit/relay/0.2.0/hop"
+	ProtoIDv2Stop = "/libp2p/circuit/relay/0.2.0/stop"
+
+	ReservationTagWeight = 10
+
+	StreamTimeout    = time.Minute
+	ConnectTimeout   = 30 * time.Second
+	HandshakeTimeout = time.Minute
+
+	maxMessageSize = 4096
+)
+
+var log = logging.Logger("relay")
+
+// Relay is the (limited) relay service object.
+type Relay struct {
+	ctx    context.Context
+	cancel func()
+
+	host host.Host
+	rc   Resources
+	acl  ACLFilter
+
+	mx    sync.Mutex
+	rsvp  map[peer.ID]time.Time
+	conns map[peer.ID]int
+}
+
+// New constructs a new limited relay that can provide relay services in the given host.
+func New(h host.Host, opts ...Option) (*Relay, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	r := &Relay{
+		ctx:    ctx,
+		cancel: cancel,
+		host:   h,
+		rc:     DefaultResources(),
+		acl:    nil,
+		rsvp:   make(map[peer.ID]time.Time),
+		conns:  make(map[peer.ID]int),
+	}
+
+	for _, opt := range opts {
+		err := opt(r)
+		if err != nil {
+			return nil, fmt.Errorf("error applying relay option: %w", err)
+		}
+	}
+
+	h.SetStreamHandler(ProtoIDv2Hop, r.handleStream)
+	h.Network().Notify(
+		&network.NotifyBundle{
+			DisconnectedF: r.disconnected,
+		})
+	go r.background()
+
+	return r, nil
+}
+
+func (r *Relay) Close() error {
```

Comment:
Let's cancel the context and wait till the go-routine finishes to ensure that we've wrapped up everything before we return from `Close`.

Best to use a `sync.Once` here to ensure idempotency.

Reply:
we can't wait for everything to finish, there is no good mechanism to do that -- and we might have unlimited relay streams that never finish.

but fine on making it idempotent, an atomic variable will do.

Reply:
so no on making it wait, yes on using a better idempotency mechanism that can support concurrent closes without fear of panic.